src/util/pg.py

```diff
@@ -608,18 +608,45 @@ def view_exists(cr, view):
     return bool(cr.rowcount)
 
 
-def remove_constraint(cr, table, name, cascade=False):
+def remove_constraint(cr, table, name, cascade=False, warn=True):
+    """
+    Remove a table constraint.
+
+    This function removes the constraint `name` from the table named `table`. It also
+    removes records from `ir_model_constraint` and its xmlids. It logs not found
+    constraints.
+
+    .. note::
+
+       If there is no constraint `name`, this function will attempt to remove
+       ``{table}_{name}``, the latter is the default name the ORM uses for constraints
+       created from `_sql_constraints`. This name mismatch will be logged.
+
+    :param str table: table from where to remove the constraint
+    :param str name: name of the constraint to remove
+    :param bool cascade: cascade the constraint removal
+    :param bool warn: use warning level when logging not found constraints, otherwise use
+                      info level
+    :return: whether the constraint was removed
+    :rtype: bool
+    """
     _validate_table(table)
+    log = _logger.warning if warn else _logger.info
     cascade = "CASCADE" if cascade else ""
     cr.execute('ALTER TABLE "{}" DROP CONSTRAINT IF EXISTS "{}" {}'.format(table, name, cascade))
-    # small exception: odoo specific action.
-    # needs to be kept here to avoid resurive imports.
-    # a solution would be to not do it now but adds an `end-` script that remove the invalid entries
-    # from the `ir_model_constraint` table
+    # Exceptionally remove Odoo records, even if we are in PG land on this file. This is somehow
+    # valid because ir.model.constraint are ORM low-level objects that relate directly to table
+    # constraints.
     cr.execute("DELETE FROM ir_model_constraint WHERE name = %s RETURNING id", [name])
-    if cr.rowcount:
-        ids = tuple(c for (c,) in cr.fetchall())
-        cr.execute("DELETE FROM ir_model_data WHERE model = 'ir.model.constraint' AND res_id IN %s", [ids])
+    if not cr.rowcount:
+        if name.startswith(table + "_"):
+            log("%r not found in ir_model_constraint, table=%r", name, table)
+            return False
+        log("%r not found in ir_model_constraint, attempting to remove with table %r prefix", name, table)
+        return remove_constraint(cr, table, "{}_{}".format(table, name), cascade, warn)
+    ids = tuple(c for (c,) in cr.fetchall())
+    cr.execute("DELETE FROM ir_model_data WHERE model = 'ir.model.constraint' AND res_id IN %s", [ids])
+    return True
```

Comment:
```suggestion
    if cr.rowcount:
        ids = tuple(c for (c,) in cr.fetchall())
        cr.execute("DELETE FROM ir_model_data WHERE model = 'ir.model.constraint' AND res_id IN %s", [ids])
        return True
    if name.startswith(table + "_"):
        log("%r not found in ir_model_constraint, table=%r", name, table)
        return False
    log("%r not found in ir_model_constraint, attempting to remove with table %r prefix", name, table)
    return remove_constraint(cr, table, "{}_{}".format(table, name), cascade, warn)
```

Reply:
Arguments:

* Avoid changing the blame for the normal case
* No nested `if`s
* `fetchall()` right after query

Reply:
Indeed, this code is a remnant of a non-recursive version I had first...